File: src/javascript/algorithms/simulation/robotReturnToOrigin/robotReturnToOrigin.js
// Robot Return to Origin (LC 657)
// TC: O(n)  SC: O(1)

/**
 * @param {string} moves
 * @return {boolean}
 */
function judgeCircle(moves) {
    let x = 0, y = 0;
    for (const ch of moves) {
        if      (ch === 'U') y++;
        else if (ch === 'D') y--;
        else if (ch === 'L') x--;
        else if (ch === 'R') x++;
    }
    return x === 0 && y === 0;
}

function runTests() {
    const tests = [
        { moves: "UD",          expected: true  },
        { moves: "LL",          expected: false },
        { moves: "LRUD",        expected: true  },
        { moves: "LDRRUUUDDL",  expected: true  },
        { moves: "",            expected: true  },
        { moves: "UDLR",        expected: true  },
    ];

    for (const { moves, expected } of tests) {
        const result = judgeCircle(moves);
        const pass = result === expected;
        console.log(`[${pass ? "PASS" : "FAIL"}] moves="${moves}" result=${result} expected=${expected}`);
    }
}

runTests();

File: src/python/algorithms/simulation/robotReturnToOrigin/robot_return_to_origin.py
# Robot Return to Origin (LC 657)
# TC: O(n)  SC: O(1)


def judge_circle(moves: str) -> bool:
    x = y = 0
    for ch in moves:
        if ch == 'U':   y += 1
        elif ch == 'D': y -= 1
        elif ch == 'L': x -= 1
        elif ch == 'R': x += 1
    return x == 0 and y == 0


if __name__ == "__main__":
    test_cases = [
        ("UD",          True),
        ("LL",          False),
        ("LRUD",        True),
        ("LDRRUUUDDL",  True),
        ("",            True),
        ("UDLR",        True),
    ]

    for moves, expected in test_cases:
        result = judge_circle(moves)
        status = "PASS" if result == expected else "FAIL"
        print(f'[{status}] moves="{moves}" result={result} expected={expected}')
